src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,332 @@
+    /// actual size of the allocated block, which is greater than or equal to `layout.size()`.
+    ///
+    /// If this method returns an `Ok(addr)`, then the `addr` returned will be non-null address
+    /// pointing to a block of storage suitable for holding an instance represented by `layout`.
+    ///
+    /// The returned block of storage is initialized like specified by [`init`].
+    ///
+    /// [`init`]: self::AllocInit
+    ///
+    /// # Errors
+    ///
+    /// Returning `Err` indicates that either memory is exhausted or `layout` does not meet
+    /// allocator's size or alignment constraints.
+    ///
+    /// Implementations are encouraged to return `Err` on memory exhaustion rather than panicking or
+    /// aborting, but this is not a strict requirement. (Specifically: it is *legal* to implement
+    /// this trait atop an underlying native allocation library that aborts on memory exhaustion.)
+    ///
+    /// Clients wishing to abort computation in response to an allocation error are encouraged to
+    /// call the [`handle_alloc_error`] function, rather than directly invoking `panic!` or similar.
+    ///
+    /// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+    fn alloc(&mut self, layout: Layout, init: AllocInit) -> Result<(NonNull<u8>, usize), AllocErr>;
+
+    /// Deallocate the memory referenced by `ptr`.
+    ///
+    /// # Safety
+    ///
+    /// * `ptr` must denote a block of memory [*currently allocated*] via this allocator,
+    ///
+    /// * `layout` must [*fit*] that block of memory, and
+    ///
+    /// * the alignment of the `layout` must match the alignment used to allocate that block of
+    ///   memory.
+    ///
+    /// [*currently allocated*]: #currently-allocated-memory
+    /// [*fit*]: #memory-fitting
+    unsafe fn dealloc(&mut self, ptr: NonNull<u8>, layout: Layout);
+
+    /// Returns a pointer and the actual size of the allocated block. The pointer is suitable for
+    /// holding data described by a new layout with `layout`’s alignment and a size given by
+    /// `new_size`. To accomplish this, the allocator may extend the allocation referenced by `ptr`
+    /// to fit the new layout.
+    ///
+    /// If this returns `Ok`, then ownership of the memory block referenced by `ptr` has been
+    /// transferred to this allocator. The memory may or may not have been freed, and should be
+    /// considered unusable (unless of course it was transferred back to the caller again via the
+    /// return value of this method).
+    ///
+    /// If this method returns `Err`, then ownership of the memory block has not been transferred to
+    /// this allocator, and the contents of the memory block are unaltered.
+    ///
+    /// The behavior on how the allocator tries to shrink the memory can be specified by
+    /// [`placement`]. The returned block of storage is initialized like specified by [`init`].
```

Comment:
Wow, that's a long list! I'll work it off later, thank you for reviewing!
Do you think those edge cases should go into the enum documentation or the trait itself?

Reply:
I think they should go on the trait itself, since most of these are `grow`-specific and don't apply to `alloc`. If going for the interpretations that I favor, I'd say something like

> The first `layout.size()` bytes of memory are preserved or copied as appropriate from `ptr`, and the remaining bytes, from `layout.size()` to the returned `actual_size`, are initialized according to `init`.

And on `alloc`:

> The returned block of storage is initialized as specified by `init`, all the way up to the returned `actual_size`.